### ISSUE TITLE
crash NoMethodError: undefined method `next_token_of' for nil:NilClass

### ISSUE THREAD

**Issue by whohoho** [rodjek/puppet-lint#905]
Whoops! It looks like puppet-lint has encountered an error that it doesn't
know how to handle. Please open an issue at https://github.com/rodjek/puppet-lint
and paste the following output into the issue description.
---
puppet-lint version: 2.3.6
ruby version: 2.6.5-p114
platform: x86_64-linux
file path: ./buildfromgit.pp
file contents:
```
# defined type to build software from git
#
# @example Declaring the class
#   lalalal
#
# @repo git url
# @revision git revision
# @commands  can be a array of strings or a array of hashes
# @srcdir where to clone the repo
# @cwd default path for commands, will be created (for ex. build for cmake etc.), defaults to repo root
# @creates file that will be  created ex. /usr/local/bin/[program]
define put::buildfromgit (
#  $repo_name = $title,
  $repo_name = $name,
  String $repo = 'git://git.ardour.org/ardour/ardour.git',
  Optional[String] $revision = 'master',
  #commands can be a array of strings or a array of hashes
  $commands = ['bash waf configure', 'bash waf'],
  #$commandshash = [
  #  { cmd => 'bash waf configure', cwd => 'lala' },
  #  { cmd => 'lalalala', cwd => 'adsf' }
  #],
  String $srcdir = "/usr/local/src",
  #default path for commands, will be created (for ex. build for cmake etc.)
  Optional[String] $cwd = undef,
  Optional[String] $creates = undef,
  Optional[Array] $python3deps = undef,
  Optional[Array] $debiandeps = undef,
  Optional[Array] $debianbuilddeps = undef,
  Optional[Array] $python2deps = undef,
  Optional[Array] $nodedeps = undef,
  ){



  $repo_dir = "${srcdir}/${repo_name}"

  # When to rebuild:
  #   * when upstream git has changes
  #   * when previous build failed (any of the commands failed, or creates file not created)

  #$shouldwebuild = 

#  file { "/tmp/${repo_name}_facts.yaml":
#    #content => inline_template(' scope.to_hash.reject { |k,v| !( k.is_a?(String) && v.is_a?(String) ) }.to_yaml '),
#    content  => inline_template("<%= scope.to_hash.reject { |k,v| k.to_s =~ /(lalala|free)/ }.to_yaml %>"),
#  }

  vcsrepo { "${srcdir}/${repo_name}":
    ensure   => latest,
    provider => git,
    revision => "${revision}",
    source   => $repo,
  }


  #create build directory or somethign
  if $cwd != undef {

    $wd = "$repo_dir/$cwd"

    file { "$repo_dir/$cwd":
      ensure => directory
    }
  } else {
    $wd = $repo_dir
  }

  Exec {
    user    => 'root',
    cwd     => $wd,
    timeout => 9999,
    path    => "/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:${srcdir}:${wd}",
    environment => ['HOME=/root' ]
  }

  if $debianbuilddeps != undef {
    each($debianbuilddeps) | $pkg| {
      exec { "debianbuilddep: $pkg":
        command => "apt-get -y build-dep $pkg"
      }
    }
  }

  if $python3deps != undef {
    each($python3deps) | $pkg| {
      package { "python3-$pkg":
        name      => $pkg,
        ensure    => installed,
        provider  => 'pip3',
      }
    }
  }
  if $python2deps != undef {
    each($python2deps) | $pkg| {
      package { "python2-$pkg":
        name      => $pkg,
        ensure   => installed,
        provider => 'pip',
      }
    }
  }
  

  if $debiandeps != undef {
    ensure_packages( $debiandeps,
    { 
      ensure   => installed,
    })
  }
  if $nodedeps != undef {
    ensure_packages( $nodepkgs,
    {
      ensure   => installed,
      provider => 'npm',
      install_options => [ '-g' ],
    })
  }



  #write a file with puppet run id
  exec {"${repo_name} echo ${transaction_uuid} > ${srcdir}/current_puppet_run":
    command => "echo ${transaction_uuid} > ${srcdir}/current_puppet_run",  
  }

  #write a file with puppet run id when there is changes in the repo
  exec {"${srcdir}/${repo_name}.haschanged":
    subscribe   => Vcsrepo["${srcdir}/${repo_name}"],
    refreshonly => true,
    command => "echo ${transaction_uuid} > ${srcdir}/${repo_name}.haschanged"  
  }

  $prevcmd = "${srcdir}/${repo_name}.haschanged"
  exec {"${name}: shouldrebuild":
      command => "echo rebuild",
      #subscribe   => Vcsrepo["${srcdir}/${repo_name}"],
      #refreshonly => true,
      #onlyif "cat ${srcdir}/{$repo_name}.haschanged -eq ${repo_name} OR ${creates} does not exist} "
      onlyif => "cmp ${srcdir}/current_puppet_run} ${srcdir}/${repo_name}.haschanged"
    }

  $cmddefaults = { 
    cmd => '/bin/false', 
    cwd => $wd, 
    something => 'lala'
    }

  each($commands) | $command| {

    if $command =~ String {
      $cmdhash = $cmddefaults + { cmd => $command }
    }
    elsif $command =~ Hash {
      $cmdhash = $cmddefaults + $cmd
    }
    else {
      fail("$command is not a string or a hash")
    }
    
    #depend on previous 
    #->  
  
    exec {"${name}: ${command}":
      command => $cmdhash[cmd],
      cwd     => $cmdhash[cwd],
      #subscribe   => Vcsrepo["${srcdir}/${repo_name}"],
      #refreshonly => true,
      #onlyif "cat ${srcdir}/{$repo_name}.haschanged -eq ${repo_name} OR ${creates} does not exist} ",
      #onlyif => "cmp ${srcdir}/current_puppet_run} ${srcdir}/${repo_name}.haschanged",
      # FIXME: how to do this??
      require => Exec[$prevcmd],
    }
    #FIXME: how to do this??
    $prevcmd = "${name}: ${command}"
  
  }

   exec {"${name}: buildsuccess":
      command => "echo ${transaction_uuid} > ${srcdir}/${repo_name}.success",
      #subscribe   => Vcsrepo["${srcdir}/${repo_name}"],
      #refreshonly => true,
      #onlyif "cat ${srcdir}/{$repo_name}.haschanged -eq ${repo_name} OR ${creates} does not exist} "
      onlyif => "cmp ${srcdir}/current_puppet_run} ${srcdir}/${repo_name}.haschanged"
    }

 



}



```
error:
```
NoMethodError: undefined method `next_token_of' for nil:NilClass
/nix/store/7ns7cbbvgx512dqi4p9g9pgas5lnmmj9-ruby2.6.5-puppet-lint-2.3.6/lib/ruby/gems/2.6.0/gems/puppet-lint-2.3.6/lib/puppet-lint/plugins/check_resources/ensure_first_param.rb:32:in `fix'
/nix/store/7ns7cbbvgx512dqi4p9g9pgas5lnmmj9-ruby2.6.5-puppet-lint-2.3.6/lib/ruby/gems/2.6.0/gems/puppet-lint-2.3.6/lib/puppet-lint/checkplugin.rb:42:in `block in fix_problems'
/nix/store/7ns7cbbvgx512dqi4p9g9pgas5lnmmj9-ruby2.6.5-puppet-lint-2.3.6/lib/ruby/gems/2.6.0/gems/puppet-lint-2.3.6/lib/puppet-lint/checkplugin.rb:38:in `each'
/nix/store/7ns7cbbvgx512dqi4p9g9pgas5lnmmj9-ruby2.6.5-puppet-lint-2.3.6/lib/ruby/gems/2.6.0/gems/puppet-lint-2.3.6/lib/puppet-lint/checkplugin.rb:38:in `fix_problems'
/nix/store/7ns7cbbvgx512dqi4p9g9pgas5lnmmj9-ruby2.6.5-puppet-lint-2.3.6/lib/ruby/gems/2.6.0/gems/puppet-lint-2.3.6/lib/puppet-lint/checks.rb:67:in `block in run'
/nix/store/7ns7cbbvgx512dqi4p9g9pgas5lnmmj9-ruby2.6.5-puppet-lint-2.3.6/lib/ruby/gems/2.6.0/gems/puppet-lint-2.3.6/lib/puppet-lint/checks.rb:65:in `each'
/nix/store/7ns7cbbvgx512dqi4p9g9pgas5lnmmj9-ruby2.6.5-puppet-lint-2.3.6/lib/ruby/gems/2.6.0/gems/puppet-lint-2.3.6/lib/puppet-lint/checks.rb:65:in `run'
/nix/store/7ns7cbbvgx512dqi4p9g9pgas5lnmmj9-ruby2.6.5-puppet-lint-2.3.6/lib/ruby/gems/2.6.0/gems/puppet-lint-2.3.6/lib/puppet-lint.rb:205:in `run'
/nix/store/7ns7cbbvgx512dqi4p9g9pgas5lnmmj9-ruby2.6.5-puppet-lint-2.3.6/lib/ruby/gems/2.6.0/gems/puppet-lint-2.3.6/lib/puppet-lint/bin.rb:66:in `block in run'
/nix/store/7ns7cbbvgx512dqi4p9g9pgas5lnmmj9-ruby2.6.5-puppet-lint-2.3.6/lib/ruby/gems/2.6.0/gems/puppet-lint-2.3.6/lib/puppet-lint/bin.rb:62:in `each'
/nix/store/7ns7cbbvgx512dqi4p9g9pgas5lnmmj9-ruby2.6.5-puppet-lint-2.3.6/lib/ruby/gems/2.6.0/gems/puppet-lint-2.3.6/lib/puppet-lint/bin.rb:62:in `run'
/nix/store/9ihaf09wax4sbz6sm99dnwji15bhhpzs-puppet-lint-2.3.6/lib/ruby/gems/2.6.0/gems/puppet-lint-2.3.6/bin/puppet-lint:7:in `<top (required)>'
/home/sysadm/.nix-profile/bin/puppet-lint:18:in `load'
/home/sysadm/.nix-profile/bin/puppet-lint:18:in `<main>'

**Comment by whohoho** [comment]
and the same on another file:

Whoops! It looks like puppet-lint has encountered an error that it doesn't
know how to handle. Please open an issue at https://github.com/rodjek/puppet-lint
and paste the following output into the issue description.
---
puppet-lint version: 2.3.6
ruby version: 2.6.5-p114
platform: x86_64-linux
file path: puppet6.pp
file contents:
```
class put::puppet6 {

  #READ this: https://www.example42.com/2018/10/08/puppet6-ca-upgrading/


  # puppetlabs key
  apt::key { 'puppet gpg key':
    id     => '6F6B15509CF8E59E6E469F327F438280EF8D349F',
    server => 'pgp.mit.edu',
  }


  $gpgkey = @("EOF")
-----BEGIN PGP PUBLIC KEY BLOCK-----

mQINBFe2Iz4BEADqbv/nWmR26bsivTDOLqrfBEvRu9kSfDMzYh9Bmik1A8Z036Eg
h5+TZD8Rrd5TErLQ6eZFmQXk9yKFoa9/C4aBjmsL/u0yeMmVb7/66i+x3eAYGLzV
FyunArjtefZyxq0B2mdRHE8kwl5XGl8015T5RGHCTEhpX14O9yigI7gtliRoZcl3
hfXtedcvweOf9VrV+t5LF4PrZejom8VcB5CE2pdQ+23KZD48+Cx/sHSLHDtahOTQ
5HgwOLK7rBll8djFgIqP/UvhOqnZGIsg4MzTvWd/vwanocfY8BPwwodpX6rPUrD2
aXPsaPeM3Q0juDnJT03c4i0jwCoYPg865sqBBrpOQyefxWD6UzGKYkZbaKeobrTB
xUKUlaz5agSK12j4N+cqVuZUBAWcokXLRrcftt55B8jz/Mwhx8kl6Qtrnzco9tBG
T5JN5vXMkETDjN/TqfB0D0OsLTYOp3jj4hpMpG377Q+6D71YuwfAsikfnpUtEBxe
NixXuKAIqrgG8trfODV+yYYWzfdM2vuuYiZW9pGAdm8ao+JalDZss3HL7oVYXSJp
MIjjhi78beuNflkdL76ACy81t2TvpxoPoUIG098kW3xd720oqQkyWJTgM+wV96bD
ycmRgNQpvqHYKWtZIyZCTzKzTTIdqg/sbE/D8cHGmoy0eHUDshcE0EtxsQARAQAB
tEhQdXBwZXQsIEluYy4gUmVsZWFzZSBLZXkgKFB1cHBldCwgSW5jLiBSZWxlYXNl
IEtleSkgPHJlbGVhc2VAcHVwcGV0LmNvbT6JAj4EEwECACgFAle2Iz4CGwMFCQlm
AYAGCwkIBwMCBhUIAgkKCwQWAgMBAh4BAheAAAoJEH9DgoDvjTSfIN0P/jcCRzK8
WIdhcNz5dkj7xRZb8Oft2yDfenQmzb1SwGGa96IwJFcjF4Nq7ymcDUqunS2DEDb2
gCucsqmW1ubkaggsYbc9voz/SQwhsQpBjfWbuyOX9DWmW6av/aB1F85wP79gyfqT
uidTGxQE6EhDbLe7tuvxOHfM1bKsUtI+0n9TALLLHfXUEdtaXCwMlJuO1IIn1PWa
H7HzyEjw6OW/cy73oM9nuErBIio1O60slPLOW2XNhdWZJCRWkcXyuumRjoepz7WN
1JgsLOTcB7rcQaBP3pDN0O/Om5dlDQ6oYitoJs/F0gfEgwK68Uy8k8sUR+FLLJqM
o0CwOg6CeWU4ShAEd1xZxVYW6VOOKlz9x9dvjIVDn2SlTBDmLS99ySlQS57rjGPf
GwlRUnuZP4OeSuoFNNJNb9PO6XFSP66eNHFbEpIoBU7phBzwWpTXNsW+kAcY8Rno
8GzKR/2FRsxe5Nhfh8xy88U7BA0tqxWdqpk/ym+wDcgHBfSRt0dPFnbaHAiMRlgX
J/NPHBQtkoEdQTKA+ICxcNTUMvsPDQgZcU1/ViLMN+6kZaGNDVcPeMgDvqxu0e/T
b3uYiId38HYbHmD6rDrOQL/2VPPXbdGbxDGQUgX1DfdOuFXw1hSTilwI1KdXxUXD
sCsZbchgliqGcI1l2En62+6pI2x5XQqqiJ7+
=HpaX
-----END PGP PUBLIC KEY BLOCK-----
EOF

  #copy puppet.conf
  $puppetconf = @("EOF")
[main]
    ssldir = /var/lib/puppet/ssl
    vardir = /var/lib/puppet
    certname = ${fqdn}
    server = ${puppet_master_server}

[agent]
runinterval=48h
EOF

  # for some reason, puppet fetching key from keyserver is not working....
  file { '/root/DEB-GPG-KEY-puppet':
    content => $gpgkey,
  }
  -> exec { 'manually import pup gpg key':
    command => 'apt-key add /root/DEB-GPG-KEY-puppet',
    path    => '/usr/bin:/usr/sbin:/bin:/usr/local/bin:/usr/local/sbin:/sbin',
  }
  -> apt::key { 'puppet6 gpg key':
    id     => '6F6B15509CF8E59E6E469F327F438280EF8D349F',
    server => 'pgp.mit.edu',
  }
  -> apt::source { 'puppet6':
    comment  => 'puppet6',
    location => 'http://apt.puppetlabs.com',
    repos    => 'puppet6',
   #FIXME: change to lsbdistcodename when buster puppet is released
    release  => $lsbdistcodename,
    include  => {
#      'src' => true,
      'deb' => true,
    },
  }
  -> exec { 'apt-update pupp6':
    command => '/usr/bin/apt-get update',
    path    => '/usr/bin:/usr/sbin:/bin:/usr/local/bin:/usr/local/sbin:/sbin',
    #schedule => 'everyday'
  }
  -> file { '/etc/puppetlabs':
    ensure => directory,
  }
  -> file { '/etc/puppetlabs/puppet':
    ensure => directory,
  }
  -> file { '/etc/puppetlabs/puppet/puppet.conf':
    content => $puppetconf,
  }
  -> package{ 'puppet6-release':
    ensure   => installed,
    provider => apt,
  }
  -> package{ 'puppet':
    ensure   => absent,
    provider => apt,
  }
  -> package{ 'puppet-agent':
    provider => ensure   => installed,
    apt,
  }

}


```
error:
```
NoMethodError: undefined method `next_token_of' for nil:NilClass
/nix/store/7ns7cbbvgx512dqi4p9g9pgas5lnmmj9-ruby2.6.5-puppet-lint-2.3.6/lib/ruby/gems/2.6.0/gems/puppet-lint-2.3.6/lib/puppet-lint/plugins/check_resources/ensure_first_param.rb:32:in `fix'
/nix/store/7ns7cbbvgx512dqi4p9g9pgas5lnmmj9-ruby2.6.5-puppet-lint-2.3.6/lib/ruby/gems/2.6.0/gems/puppet-lint-2.3.6/lib/puppet-lint/checkplugin.rb:42:in `block in fix_problems'
/nix/store/7ns7cbbvgx512dqi4p9g9pgas5lnmmj9-ruby2.6.5-puppet-lint-2.3.6/lib/ruby/gems/2.6.0/gems/puppet-lint-2.3.6/lib/puppet-lint/checkplugin.rb:38:in `each'
/nix/store/7ns7cbbvgx512dqi4p9g9pgas5lnmmj9-ruby2.6.5-puppet-lint-2.3.6/lib/ruby/gems/2.6.0/gems/puppet-lint-2.3.6/lib/puppet-lint/checkplugin.rb:38:in `fix_problems'
/nix/store/7ns7cbbvgx512dqi4p9g9pgas5lnmmj9-ruby2.6.5-puppet-lint-2.3.6/lib/ruby/gems/2.6.0/gems/puppet-lint-2.3.6/lib/puppet-lint/checks.rb:67:in `block in run'
/nix/store/7ns7cbbvgx512dqi4p9g9pgas5lnmmj9-ruby2.6.5-puppet-lint-2.3.6/lib/ruby/gems/2.6.0/gems/puppet-lint-2.3.6/lib/puppet-lint/checks.rb:65:in `each'
/nix/store/7ns7cbbvgx512dqi4p9g9pgas5lnmmj9-ruby2.6.5-puppet-lint-2.3.6/lib/ruby/gems/2.6.0/gems/puppet-lint-2.3.6/lib/puppet-lint/checks.rb:65:in `run'
/nix/store/7ns7cbbvgx512dqi4p9g9pgas5lnmmj9-ruby2.6.5-puppet-lint-2.3.6/lib/ruby/gems/2.6.0/gems/puppet-lint-2.3.6/lib/puppet-lint.rb:205:in `run'
/nix/store/7ns7cbbvgx512dqi4p9g9pgas5lnmmj9-ruby2.6.5-puppet-lint-2.3.6/lib/ruby/gems/2.6.0/gems/puppet-lint-2.3.6/lib/puppet-lint/bin.rb:66:in `block in run'
/nix/store/7ns7cbbvgx512dqi4p9g9pgas5lnmmj9-ruby2.6.5-puppet-lint-2.3.6/lib/ruby/gems/2.6.0/gems/puppet-lint-2.3.6/lib/puppet-lint/bin.rb:62:in `each'
/nix/store/7ns7cbbvgx512dqi4p9g9pgas5lnmmj9-ruby2.6.5-puppet-lint-2.3.6/lib/ruby/gems/2.6.0/gems/puppet-lint-2.3.6/lib/puppet-lint/bin.rb:62:in `run'
/nix/store/9ihaf09wax4sbz6sm99dnwji15bhhpzs-puppet-lint-2.3.6/lib/ruby/gems/2.6.0/gems/puppet-lint-2.3.6/bin/puppet-lint:7:in `<top (required)>'
/home/sysadm/.nix-profile/bin/puppet-lint:18:in `load'
/home/sysadm/.nix-profile/bin/puppet-lint:18:in `<main>'

**Comment by usev6** [comment]
Hi @whohoho, I'm able to reproduce the second problem you posted (using puppet-lint 2.3.6). The internal error goes away with the latest and greatest (puppet-lint 2.4.2). A couple of parsing errors have been fixed between 2.3.6 and 2.4.2. Could you make another try with puppet-lint 2.4.2?

For unknown reasons I don't see a crash with your first example. I've tried with the same settings as you (puppet-lint 2.3.6, ruby 2.6.5-p114, linux x86_64), but to no avail. However, maybe the crash goes away with 2.4.2 for you, too?

**Comment by rodjek** [comment]
Confirmed that the first case is the same as #906, so that'll get fixed up shortly. The second case has a syntax error in Package[puppet-agent] which as @usev6 pointed out will not throw an unhandled error in the latest version of puppet-lint, though it will not attempt to fix the lint problems until the syntax error is cleared up.
